src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -0,0 +1,731 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+
+#include <ripple/nodestore/impl/DatabaseShardImp.h>
+#include <ripple/app/ledger/InboundLedgers.h>
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/basics/random.h>
+#include <ripple/nodestore/Manager.h>
+#include <ripple/protocol/HashPrefix.h>
+
+namespace ripple {
+namespace NodeStore {
+
+DatabaseShardImp::DatabaseShardImp(Application& app,
+    std::string const& name, Stoppable& parent, Scheduler& scheduler,
+        int readThreads, Section const& config, beast::Journal j)
+    : DatabaseShard(name, parent, scheduler, readThreads, j)
+    , app_(app)
+    , config_(config)
+    , dir_(get<std::string>(config, "path"))
+    , maxDiskSpace_(get<std::uint64_t>(config, "max_size_gb") << 30)
+{
+}
+
+DatabaseShardImp::~DatabaseShardImp()
+{
+    stopThreads();
+}
+
+bool
+DatabaseShardImp::init()
+{
+    using namespace boost::filesystem;
+    {
+        // Validate backend
+        Factory* f;
+        {
+            std::string const type(get<std::string>(config_, "type"));
+            if (type.empty() || !(f = Manager::instance().find(type)))
+            {
+                JLOG(j_.error()) <<
+                    "Invalid shard store type specified";
+                return false;
+            }
+        }
+
+        std::string tmpDir;
+        auto i = 0;
+        do
+            tmpDir = "TMP" + std::to_string(i++);
+        while (is_directory(dir_ / tmpDir));
+
+        auto d = dir_ / tmpDir;
+        auto config {config_};
+        config.set("path", d.string());
+        {
+            fdLimit_ = f->createInstance(NodeObject::keyBytes,
+                config, scheduler_, j_)->fdlimit();
+            backed_ = static_cast<bool>(fdLimit_);
+        }
+        remove_all(d);
+    }
+    if (!backed_)
+        return true;
+
+    // Find shards
+    for (auto const& d : directory_iterator(dir_))
+    {
+        if (!is_directory(d))
+            continue;
+        auto dirName = d.path().stem().string();
+        if (!std::all_of(dirName.begin(), dirName.end(), ::isdigit))
+            continue;
+        auto const shardIndex {std::stoul(dirName)};
+        if (shardIndex < detail::genesisShardIndex)
+            continue;
+        auto shard = std::make_unique<Shard>(
+            shardIndex, cacheSz_, cacheAge_, j_);
+        if (!shard->open(config_, scheduler_, dir_))
+            return false;
+        usedDiskSpace_ += shard->fileSize();
+        if (shard->complete())
+            complete_.emplace(shard->index(), std::move(shard));
+        else
+        {
+            if (incomplete_)
+            {
+                JLOG(j_.error()) <<
+                    "More than one control file found";
```

Comment:
Is this equivalent to saying more than one incomplete shard found?

Reply:
There can be a maximum of one control file at any given time or none if not acquiring. There are various reasons why this error could be encountered. The user could incorrectly copy merged two shard stores or copied an incomplete shard or the system failed to remove the control file when the shard completed (very unlikely). In this unlikely event, the user will need to delete the shard(s) in question or the entire store.